### PR TITLE
Fix regression environment

### PR DESCRIPTION
**Pull request by villander**
cc: @kellyselden @rwjblue

### CODE REVIEW

**Comment by mike183** [comment]
@lennyburdette: You may also want to check this doesn't break the work you did in #529

**Comment by lennyburdette** [comment]
@mike183 I think we can go with this change instead of mine. They accomplish the same thing.

**Comment by lennyburdette** [comment]
Oh wait, mine was merged already. I’ll check that it doesn’t break our app again. Thanks for the heads up!

**Comment by villander** [comment]
@lennyburdette your solution had a hole, where the value of environment, was always `development` or `undefined`. Now it's all right, I imagine. Test your app and let me know, please.

File: node-tests/acceptance/build-test.js [hunk 1]
```diff
@@ -73,10 +73,7 @@ describe('Acceptance', function() {
         expect(output.manifest()).to.deep.equal(expectedManifests['lazy']);
         output.contains('assets/node-asset-manifest.js');
         // Engine inherites its environment from the parent
-        output.contains(
-          `engines-dist/${engineName}/config/environment.js`,
-          /"environment":"development"/
```
Comment (rwjblue) [inline node-tests/acceptance/build-test.js]: Why would we want this assertion removed?

Comment (villander) [reply node-tests/acceptance/build-test.js]: @rwjblue the old solution has a problem where the `environment` always returns the` development` value. And that test just covered that scenario, I preferred to make sure that only the `config/environment` is exposed out.

But I agree we should test all scenarios.

**Comment by villander** [comment]
> We need to change the way we cache here, specifically we need to support building two environments during a single build (this is how ember s + /tests actually works.

Can you tweak the caching to a map (or other structure that supports different configs per-environment)?

@rwjblue you can show me some example of how `ember-cli` do this?

**Comment by rwjblue** [comment]
[Here](https://github.com/ember-cli/ember-cli/blob/ea4e7758c37419a257d317c37581f901d8e63ced/lib/models/project.js#L222-L257) is where ember-cli manages the config map.

**Comment by stefanpenner** [comment]
I believe @rwjblue  is recommending something like:

```
diff --git a/lib/engine-addon.js b/lib/engine-addon.js
index c15e051..faf8bf5 100644
--- a/lib/engine-addon.js
+++ b/lib/engine-addon.js
@@ -964,13 +964,19 @@ module.exports = {
       @return {Object} Configuration object that will be provided to the engine.
     */
     options.engineConfig = function() {
-      if (this._engineConfig) {
-        return this._engineConfig;
-      }
-
       var host = findHost.call(this);
       var env = host.env;
+      let key = `${host.project.configPath()}|${env}`;
+      if (!this._engineConfig) {
+        this._engineConfig = new Map();
+      }
+
+      if (this._engineConfig.has(key)) {
+        this._engineConfig.get(key);
+      }
+
       var baseConfig = {};
+      var config;
 
       var configPath = 'config';
 
@@ -990,10 +996,13 @@ module.exports = {
 
         var addonsConfig = this.getAddonsConfig(env, engineConfig);
 
-        return (this._engineConfig = merge(addonsConfig, engineConfig));
+        config = merge(addonsConfig, engineConfig);
       } else {
-        return (this._engineConfig = this.getAddonsConfig(env, {}));
+        config =  this.getAddonsConfig(env, {});
       }
+
+      this._engineConfig.set(key, config);
+      return config;
     };
 
     /**
@@ -1166,6 +1175,7 @@ module.exports = {
 
       return mergedTree;
     };
+
     return options;
   },
 };

```

File: node-tests/acceptance/build-test.js (hunk 1 above)
Comment (stefanpenner) [reply node-tests/acceptance/build-test.js]: yes, we will need to test more then one scenario

Comment (villander) [reply node-tests/acceptance/build-test.js]: I'll do this between today and tomorrow.

**Comment by stefanpenner** [comment]
crap.... i pushed the wrong branch when I tried to rebase....

**Comment by stefanpenner** [comment]
since i screwed this up, I'll take it to completion: new branch/pr -> https://github.com/ember-engines/ember-engines/pull/545

**Comment by villander** [comment]
@stefanpenner I was working on this now haha

I'll leave you finish this =D

**Comment by stefanpenner** [comment]
@villander oh geeze, I am sorry. I attempted to merely help with the rebase, but I managed to fat finger this one..

**Comment by villander** [comment]
No problem @stefanpenner. I'll learn from you =D

**Pull request closed without merging**